scripts/build_test.py

```diff
@@ -14,29 +14,270 @@
+        self.assertTrue(
+            'Hash from file named %s does not match hash dict values'
+            % hashed_base_filename in incorrectHashInFilename.exception)
+
+    def test_process_html(self):
+        """Test process_html to remove whitespaces and hash filepaths."""
+        base_source_path = os.path.join(
+            build.TEMPLATES_CORE_DIRNAMES_TO_DIRPATHS['dev_dir'], 'pages',
+            'base.html')
+        # Prepare a file_stream object from StringIO.
+        minified_html_file_stream = StringIO.StringIO()
+        # Obtain actual file hashes of /templates to add hash to all filepaths
+        # within the HTML file. The end result will look like:
+        # E.g <script ... app.js></script>
+        # --> <script ... app.[hash].js></script>.
+        file_hashes = build.get_file_hashes(
+            build.TEMPLATES_CORE_DIRNAMES_TO_DIRPATHS['dev_dir'])
+        # Reading from actual base.html file and not simply a dummy file.
+        build._ensure_files_exist([base_source_path])
+        # Assert that base.html has white spaces.
+
+        with open(base_source_path, 'r') as source_base_file:
+            source_base_file_content = source_base_file.read()
+            self.assertRegexpMatches(
+                source_base_file_content, r'\s{2,}',
+                msg="No white spaces detected in %s unexpectedly"
+                % base_source_path)
+            # Create a list of un-hashed filenames in base.html.
+            base_file_unhashed_filenames = []
+            for filepath, _ in file_hashes.iteritems():
+                # Not hashing HTML files.
+                if filepath.endswith('.html'):
+                    continue
+                result = re.findall(filepath, source_base_file_content)
+                if result:
+                    base_file_unhashed_filenames.append(result[0])
+            self.assertGreater(len(base_file_unhashed_filenames), 0)
+
+            build.process_html(
+                source_base_file, minified_html_file_stream, file_hashes)
+
+        minified_html_file_content = minified_html_file_stream.getvalue()
+        self.assertNotRegexpMatches(
+            minified_html_file_content, r'\s{2,}',
+            msg='All white spaces must be removed from %s' %
+            base_source_path)
+        # Assert that all filenames must be hashed.
+        for unhashed_filename in base_file_unhashed_filenames:
+            result = re.findall(
+                unhashed_filename, minified_html_file_content)
+            self.assertEqual(len(result), 0)
+
+    def test_hash_should_be_inserted(self):
+        """Test hash_should_be_inserted to return the correct boolean value
+        for filepath that should be hashed.
+        """
+        with self.swap(
+            build, 'FILEPATHS_NOT_TO_RENAME', (
+                'path/to/fonts/*', 'path/to/third_party.min.js.map',
+                'path/to/third_party.min.css.map')):
```

Comment:
Should we check python files as well? (Not sure -- it just occurred to me that they don't need a hash.)

Reply:
So `hash_should_be_inserted` is used in 2 places:
`...copy_from_source_to_target()`: This function already skips Python file before calling `hash_should_be_inserted`
`_verify_build()`: This function looks into the final directory containing built files, which must not have Python files at any point in time.
Checking for Python files would return true here. Since it is assumed , by `hash_should_be_inserted`, that all Python files are already skipped/ does not exist.

Reply:
Hm. Should this be part of the contract of the original function then?

When reviewing a unit test, my aim is to check that the tests exercise the contract as described in the function. So, we should be really explicit in the function docstring about what's valid input and what's not.

Reply:
Here is `hash_should_be_inserted` in its entirety, 
```Python
# Files with this paths should be moved to build directory but shouldn't be
# renamed (i.e. the filepath shouldn't contain hash)
# This is because these files are referenced from third party files or don't
# need cache invalidation.
FILEPATHS_NOT_TO_RENAME = (
    'third_party/generated/fonts/*',
    'third_party/generated/js/third_party.min.js.map')


def hash_should_be_inserted(filepath):
    """Returns if the file should be renamed to include hash in
    the path.

    Args:
        filepath: str. Path relative to directory we are currently building.

    Returns:
        bool. True if filepath should contain hash else False.
    """
    return not any(fnmatch.fnmatch(filepath, pattern) for pattern
                   in FILEPATHS_NOT_TO_RENAME)

```
As you can see `hash_should_be_inserted` was designed to be used only in the context of no Python files, i.e in the /build directory. Should I add the Python rules to FILEPATHS_NOT_TO_RENAME?

Reply:
Ah maybe I was unclear. When writing unit tests, it doesn't actually matter what the implementation of the original function is. The thing that matters is the interface of the function, i.e. what comes in and what goes out (plus any side-effects, if applicable).

So, as long as the docstring matches what the tests are testing, I think it's fine. I.e. you could either define the function as one that should be able to handle Python filepaths, or require in the args that the filepath doesn't end in .py. Whichever you go with, the tests should fully test that.

On the whole, I guess it's probably cleaner to just make this a general-purpose function and include python rules in FILEPATHS_NOT_TO_RENAME (you might need to make minor modifications to the second part of its comment though if you do that). But the more general point is to think through whether this covers everything else, and, if not, update the docstrings accordingly.

Reply:
Done :)